### ISSUE TITLE
Spring Boot 1.2.1 GroovyTemplates issue in IntellijIdea 14

### ISSUE THREAD

**Issue by bondarYehor** [spring-projects/spring-boot#2319]
Hello,
I've faced an issue with resolving Groovy Templates when try to run SpringBoot application : https://github.com/melix/springboot-groovytemplates

When I try to hit any url I get 404 error. When I dig deeper and debug template resolving stack I found and `IOException` in `org.springframework.web.servlet.view.groovy.GroovyMarkupConfigurer#resolveTemplate`
Seems like `groovy.text.markup.MarkupTemplateEngine$TemplateGroovyClassLoader` class loader can't allocate resources in `templates/` folder.

BTW, when I run this application with "gradle run" everything works as expected and templates resolver works great. I debug such run mode and debug stack looks similar to Idea's one.

Does anybody know what can cause such unexpected behavior?

**Comment by wilkinsona** [comment]
spring-boot-sample-web-groovy-templates has the same problem in IntelliJ

**Comment by wilkinsona** [comment]
When the template class loader's being created in IntelliJ `classpath:/templates/` resolves to `file:/Library/Java/JavaVirtualMachines/jdk1.8.0.jdk/Contents/Home/jre/lib/htmlconverter.jar!/templates/`which prevents it from finding the app's templates. This doesn't happen in Eclipse or, as observed above by @bondarYehor, when run from the command line using `./gradlew bootRun`.

**Comment by wilkinsona** [comment]
A workaround is to change the prefix that's used when looking up templates by adding the following line the `application.properties`:

    spring.groovy.template.prefix: classpath*:/templates/

Using `classpath*` means that both the location in htmlconverter.jar and `target/classes/templates` are found and queried when looking for a template.

I've yet to figure out why htmlconverter.jar is found first when running in IntelliJ, but I strongly suspect it's an IDE bug/quirk that we can do little about beyond recommending the above-described workaround.

**Comment by wilkinsona** [comment]
As suspected, an IntelliJ quirk is the cause of the problem. By default, the classpath for the module places the JDK ahead of the module's own source. Opening the module's settings (F4) and moving `<Module source>` to the top of the list also resolves the problem.

@snicoll As our resident IntelliJ user, do you have any advice? Is there a global config option that we can recommend people set to get a sane classpath by default? Is it worth mentioning this to the IntelliJ team?

**Comment by snicoll** [comment]
I already pinged @YannCebron about it. I'll do that again and report here.

**Comment by chashnikov** [comment]
Yes, IntelliJ IDEA indeed places JDK classes before the module classes in the classpath by default, and currently there is no option to change this behavior. It's possible to detect such cases and change the order of classpath items in Spring Boot plugin, but an approach where the first 'templates' package from the classpath is used to search the templates looks unreliable and fragile for me. 'templates' is a common name and even if we fix the problem with htmlconverter.jar from JDK this can be easily broken by another library placed before your jar in classpath. So I think it would be better to either search in all 'templates' packages by default or replace 'templates' name by some more specific name like 'spring-boot-templates' .

**Comment by wilkinsona** [comment]
Thanks, @chashnikov.

> an approach where the first 'templates' package from the classpath is used to search the templates looks unreliable and fragile for me.

It's not when we know that a module's own resources will be considered first. This is true in Eclipse, when run using `mvn spring-boot:run`, `gradle bootRun`, and `java -jar`.

> even if we fix the problem with htmlconverter.jar from JDK this can be easily broken by another library placed before your jar in classpath

Assuming predictable and consistent ordering of items on the classpath, that should never happen.

>  I think it would be better to either search in all 'templates' packages by default

We'd prefer not to do that as it means that every jar on the classpath needs to be checked which may hurt performance more than we'd like. Users who want to spread their templates over multiple jars have the option of using a `classpath*:` configuration, but it should be opt in as, for most people, it's unnecessary.

In summary, as I see it, the problem with IntelliJ's behaviour is that for both Maven and Gradle projects, the classpath in IntelliJ is different to the classpath in Maven or Gradle. I would expect IntelliJ to create a classpath for the module that mirrors, as closely as possible, the classpath that the module's configured build system would use. As things stand, I can run `mvn spring-boot:run` from within IntelliJ and everything works, but trying to run the main class directly leads to the problem described above. I'd argue that this inconsistency in IntelliJ's classpath configuration is what is unreliable and fragile.

**Comment by wilkinsona** [comment]
I've added a tip in the documentation to alert people about the different classpath ordering in IDEA
